Enhance angle calculations of the Random mod

Pasi4K5 opened the pull request:
I originally wanted to refrain from submitting any PRs for the Random mod because of the broken replay problem (https://github.com/ppy/osu/discussions/16048), but since it seems to be fine (https://github.com/ppy/osu/pull/17194#issuecomment-1079673187) and multiple PRs have been submitted recently, I’m now doing it anyway.
### Problem
The main problem this PR addresses is that the angles between far spaced hit objects are way too wide in many cases, which makes some patterns extremely uncomfortable and unreasonably difficult to play. Furthermore, it turns patterns that were originally jump patterns into a big mess of hit objects that goes in circles around the screen.
![image](https://user-images.githubusercontent.com/47356629/174479921-7794967d-51a8-4ac4-a78c-ac3764e6f1e8.png)
The reason for this issue is the `rateOfChangeMultiplier`, which is random multiplier between -1 and 1 that changes every 5 iterations.
![image](https://user-images.githubusercontent.com/47356629/174479941-3aa339dd-1769-4468-953c-16bd199c91c4.png)
β is angle that is stored in `positionInfo.RelativeAngle`. Before storing the angle β into `positionInfo.RelativeAngle`, it is multiplied by `rateOfChangeMultiplier`. Since the absolute value of `rateOfChangeMultiplier` is never greater than 1, β can only get smaller, which results in a wider angle.
### Solution
To resolve this issue, I replaced the `rateOfChangeMultiplier` with 3 new variables. `sequenceOffset`, `flowChangeOffset` and `oneTimeOffset`, each one fulfilling a certain purpose. Those variables are used to change angles depending on whether or not a hit object is on a (down-)beat or starts a new combo. I also used normal distribution to generate the values, hence the method `RandomGaussian()`.
### Further enhancements
So far, the relationship between the distance between 2 hit objects and the target angle was linear. As a mathematical function with α being the target angle as shown in the above figure in rad and x being the distance/spacing from previous to current object, it can be expressed like this:
![image](https://user-images.githubusercontent.com/47356629/174482343-38c17475-b06e-40e3-aa81-fc88501b1998.png)
What kind of mathematical function is best suited for calculating the target angle is certainly debatable, but after testing multiple functions and doing a little bit of parameter tweaking, I came up with this logistic function:
![image](https://user-images.githubusercontent.com/47356629/174482326-0c7a2871-97a8-4401-bfbb-312b7d8b9917.png)
Here is a comparison between a graph of the old/current function versus the one proposed in this PR:
![image](https://user-images.githubusercontent.com/47356629/181072216-129a79c2-f73b-4eff-bc2e-acadae4d22f1.png)
I implemented this function in a separate method (`GetRelativeTargetAngle()`).

Pasi4K5 committed:
Enhance target angle calculation

Pasi4K5 committed:
Adjust target angle calculation parameters

Pasi4K5 committed:
Remove unused field

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
+        public static bool IsHitObjectOnBeat(OsuBeatmap beatmap, OsuHitObject hitObject, bool downbeatsOnly = false)
+        {
+            var timingPoints = beatmap.ControlPointInfo.TimingPoints;
+            var currentTimingPoint = timingPoints.Reverse().FirstOrDefault(p => p.Time <= hitObject.StartTime);
```
```suggestion
            var currentTimingPoint = timingPoints.LastOrDefault(p => p.Time <= hitObject.StartTime);
```

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
             );
         }
+
+        public static bool IsHitObjectOnBeat(OsuBeatmap beatmap, OsuHitObject hitObject, bool downbeatsOnly = false)
```
`downbeatsOnly` needs xmldoc.

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
+                ? currentTimingPoint.BeatLength * currentTimingPoint.TimeSignature.Numerator
+                : currentTimingPoint.BeatLength;
+
+            return (timeSinceTimingPoint + 1) % length < 2;
```
- Why `timeSinceTimingPoint + 1`?
- I would strongly advise against using the `%` operator on floats. It can return results that are just not correct if unlucky with floating point representation issues. Recommend using rounded division instead.

Pasi4K5 replied:
This already contains a tolerance of +-1. With infinite precision, this would already work with returning `timeSinceTimingPoint % length == 0`, but by adding 1 and using `< 2` instead of `== 0` I made sure that no floating point issues occur. I also tested this method and I didn't find a single hit object on a (down-)beat that wasn't correctly identified by this method.

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
+
+        public static float GetRelativeTargetAngle(float targetDistance, float offset, bool flowDirection)
+        {
+            float angle = (float)(3 / (1 + 200 * Math.Pow(MathHelper.E, 0.016 * (targetDistance - 466))) + 0.45 + offset);
```
```suggestion
            float angle = (float)(3 / (1 + 200 * Math.Exp(0.016 * (targetDistance - 466))) + 0.45 + offset);
```

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
+            return (timeSinceTimingPoint + 1) % length < 2;
+        }
+
+        public static float GetRelativeTargetAngle(float targetDistance, float offset, bool flowDirection)
```
`offset` and `flowDirection` need xmldoc. Placement of method is also questionable. I don't imagine it ever getting used outside of random mod, seems too specific.

mk56-spn replied:
maybe spacing adjust? its had some of the same issues

Pasi4K5 replied:
I don't think that spacing adjust will need this, so I moved it to `OsuModRandom.cs`. It can still be moved back at a later point if needed.

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
+
+        public static float RandomGaussian(Random rng, float mean = 0, float stdDev = 1)
+        {
+            double x1 = 1 - rng.NextDouble();
+            double x2 = 1 - rng.NextDouble();
```
I get that this is the [Box-Muller transform](https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform) but why `1 -`? `rng.NextDouble()` is already distributed onto the unit interval, is it not?

Pasi4K5 replied:
`rng.NextDouble()` returns a random number in the interval `[0;1)`, but we need `(0;1]` because `log(0)` is undefined.

smoogipoo replied:
Please add a comment mentioning this.

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils.cs:
```diff
@@ -186,5 +187,37 @@ private static Vector2 rotateVector(Vector2 vector, float rotation)
+        {
+            double x1 = 1 - rng.NextDouble();
+            double x2 = 1 - rng.NextDouble();
+            double stdNormal = Math.Sqrt(-2 * Math.Log(x1)) * Math.Sin(MathHelper.TwoPi * x2);
```
```suggestion
            double stdNormal = Math.Sqrt(-2 * Math.Log(x1)) * Math.Sin(2 * Math.PI * x2);
```

Best not to use `MathHelper` (which comes from osuTK which is mostly a dead dependency walking).

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:
```diff
@@ -37,23 +35,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
+                bool invertFlow = false;
+
+                if (i == 0 ||
+                    (positionInfos[i - 1].HitObject.NewCombo && (i <= 1 || !positionInfos[i - 2].HitObject.NewCombo) && (i <= 2 || !positionInfos[i - 3].HitObject.NewCombo)) ||
```
This condition needs to be factored into helper local methods more. I could barely tell what this was doing (namely, checking if any of the three objects in the sequence start a new combo).

Pasi4K5 replied:
Instead of creating a separate method, I used `HitObject.IndexInCurrentCombo` to make it easier to understand. Let me know if you think that it's still to complex.

Pasi4K5 replied:
Nevermind, I did end up putting them into their own methods (https://github.com/ppy/osu/pull/18769/commits/8c624d3269902ea8f83e874fbfebd204c97b8206).

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:
```diff
@@ -37,23 +35,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
+                if (i == 0 ||
+                    (positionInfos[i - 1].HitObject.NewCombo && (i <= 1 || !positionInfos[i - 2].HitObject.NewCombo) && (i <= 2 || !positionInfos[i - 3].HitObject.NewCombo)) ||
+                    OsuHitObjectGenerationUtils.IsHitObjectOnBeat(osuBeatmap, positionInfos[i - 1].HitObject, true) ||
+                    (OsuHitObjectGenerationUtils.IsHitObjectOnBeat(osuBeatmap, positionInfos[i - 1].HitObject) && rng.NextDouble() < 0.25))
```
There's a lot of randomness peppered in here and below for no real discernible reason. Why was this done? How were the thresholds chosen?

Pasi4K5 replied:
I added those because IMO changing the flow on every downbeat would be too much since that's not what most beatmaps look like, so it only changes the flow and angle offsets with a certain probability. I don't really have a reason why I chose these _exact_ values, but after testing and adjusting the thresholds I found that these values generated decent results.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:
```diff
@@ -37,23 +35,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
+                    }
+
+                    if (invertFlow)
+                        flowDirection ^= true;
```
I mean... this is kind of obscure. Would prefer

```suggestion
                        flowDirection = !flowDirection;
```

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs:
```diff
@@ -37,23 +35,51 @@ public void ApplyToBeatmap(IBeatmap beatmap)
+
+                    positionInfos[i].RelativeAngle = OsuHitObjectGenerationUtils.GetRelativeTargetAngle(
+                        positionInfos[i].DistanceFromPrevious,
+                        (sequenceOffset + oneTimeOffset) * (float)Math.Sqrt(positionInfos[i].DistanceFromPrevious) +
+                        flowChangeOffset * (float)Math.Sqrt(640 - positionInfos[i].DistanceFromPrevious),
```
What is this doing?

Pasi4K5 replied:
640 is the playfield diagonal (I re-added the field `playfield_diagonal` to make this clear). I also removed the square roots (and adjusted all the standard deviations accordingly) because it doesn't make that much of a difference here and just makes things more complex.
`sequenceOffset` and `oneTimeOffset` should affect further spaced objects more and `flowChangeOffset` should mainly affect streams, that's why they are dependent on `positionInfos[i].DistanceFromPrevious`.

Pasi4K5 committed:
Add XML documentation

Pasi4K5 committed:
Code optimisation

Pasi4K5 committed:
Adjust angle offset caluclations

Pasi4K5 committed:
Fix possible exception caused by `log(0)`

Pasi4K5 committed:
Add comments and XML doc

mk56-spn commented:
I'm interested in playtesting this a bit as feedback before it gets merged. 


 Or would it be preferable to let this through as is and iterating on it later ?

mk56-spn commented:
Alright so here are my primary remaining gripe with this PR is that i think its " angle memory" for lack of a better word is too low, what i mean by this is  that whilst it is a big step up from previous iterations i still think it fails to keep any shapeliness to maps. which is especially apparent in streams, every stream gets turned into a mushy  irregular oval ish shape

![osu_2022-07-15_13-10-00-1](https://user-images.githubusercontent.com/74463310/179212520-fb76590c-72ee-428a-8b68-99a10a0eb8e8.png)

here's an example of the randomised streams on freedom dive, they become amorphous and bland making most stream maps feel samey
![osu_2022-07-15_13-12-00](https://user-images.githubusercontent.com/74463310/179212704-2f89cf04-bfbc-4060-8967-4bfef364973e.png)
(an example of one of the streams before being randomised)

i propose a combination of the following  changes to make sure maps get randomised enough to be different and a breath of fresh air but not to the point where they lose any feel of uniqueness

**implementation of full map angle memory**
what im proposing here is a system where angles of same bpm streams and jump across the map get stored and compared , if the code detects that the objects tend to be at a certain similar angle across the whole map then the randomised version should  , wherever permissible by map boundaries , do the same across the whole  map. so if one seed make 1-2 jumps more wide angle for example, it should apply it everywhere where there was previous a notable correlation in angles.

this would solve maps like freedom dive (four dimensions ) , wizard's tower (ultimate magic) , airman etc...  from becoming amorphous , making the mod much more fun

**long slider relative angling**
sometimes random makes slider ends unreasonably harder to hit by pointing sliders ends that were initially pointed at the next circle/ slider point off into nowhere, limiting the deviation angle in these cases would make the maps a lot more fun, especially on maps with long sliders.

**allow the mirroring of sliders**

unless i am mistaken sliders can not flip themselves currently, would add to the variety a bit considering how many constraints im suggesting

Pasi4K5 commented:
> what im proposing here is a system where angles of same bpm streams and jump across the map get stored and compared , if the code detects that the objects tend to be at a certain similar angle across the whole map then the randomised version should , wherever permissible by map boundaries , do the same across the whole map. so if one seed make 1-2 jumps more wide angle for example, it should apply it everywhere where there was previous a notable correlation in angles.

> this would solve maps like freedom dive (four dimensions ) , wizard's tower (ultimate magic) , airman etc... from becoming amorphous , making the mod much more fun

Not sure about this one. I think that good generation can be achieved without using that much information from the original map. I can definitely see the problem you pointed out, but I'm not sure if this is the best way to solve it. Actually I can probably get better results just by adjusting a few parameters, but I need some time for that because I can't work on it until the end of next week.

> sometimes random makes slider ends unreasonably harder to hit by pointing sliders ends that were initially pointed at the next circle/ slider point off into nowhere, limiting the deviation angle in these cases would make the maps a lot more fun, especially on maps with long sliders.

To fix this I'd have to change the way sliders are rotated, which wouldn't really fit in this PR. Besides that, I _**think**_ this issue has been present before the changes of this PR _**if I'm not wrong**_.

> unless i am mistaken sliders can not flip themselves currently, would add to the variety a bit considering how many constraints im suggesting

Since the random mod already rotates sliders, I don't see any reason why it shouldn't be able to flip them too. Again, not something for this PR though since it's not really about angle calculations.

> ps. although it is not aimed the changes in this PR expressly, random would benefit greatly in readability from not allowing sliders that previously did not overlap to not be able to do so after random is applied

Yes, I've actually thought about this too. Most randomised maps have a lot of overlaps which makes them pretty unpleasant to look at and more difficult to read.

Pasi4K5 committed:
Adjust parameters

Pasi4K5 commented:
I tried to improve the ["stream issue"](https://github.com/ppy/osu/pull/18769#issuecomment-1185450883), mainly by changing the function in `getRelativeTargetAngle()`.

Here's a comparison between the function I proposed initially when I opened the PR (`α_new1`) vs. the one I just pushed (`α_new2`).

![image](https://user-images.githubusercontent.com/47356629/181069582-d00a0131-a55a-4011-8584-3d7b8580f396.png)

Streams are now tighter, which prevents them from going around the screen and forming weird oval shapes. It's still far from perfect, but odd shapes like [this](https://user-images.githubusercontent.com/74463310/179212520-fb76590c-72ee-428a-8b68-99a10a0eb8e8.png) appear way less frequently now and you can see the algorithm generating streams with recognisable shapes like waves and circles more often.

mk56-spn commented:
> I tried to improve the ["stream issue"](https://github.com/ppy/osu/pull/18769#issuecomment-1185450883), mainly by changing the function in `getRelativeTargetAngle()`.
> 
> Here's a comparison between the function I proposed initially when I opened the PR (`α_new1`) vs. the one I just pushed (`α_new2`).
> 
> ![image](https://user-images.githubusercontent.com/47356629/181069582-d00a0131-a55a-4011-8584-3d7b8580f396.png)
> 
> Streams are now tighter, which prevents them from going around the screen and forming weird oval shapes. It's still far from perfect, but odd shapes like [this](https://user-images.githubusercontent.com/74463310/179212520-fb76590c-72ee-428a-8b68-99a10a0eb8e8.png) appear way less frequently now and you can see the algorithm generating streams with recognisable shapes like waves and circles more often.

Gave it a try. Def feels much more Human, step in the right direction

Pasi4K5 committed:
Merge branch 'ppy:master' into enhance-randomisation

Pasi4K5 committed:
Adjust parameters

Pasi4K5 committed:
Add comments and improve code readability

Pasi4K5 commented:
After another round of tweaking standard deviations and probabilities, streams are now a bit "less random" (and thus less likely to look like [this](https://user-images.githubusercontent.com/74463310/179212520-fb76590c-72ee-428a-8b68-99a10a0eb8e8.png)) and change their flow direction more often, which prevents streams that just go in circles from appearing too often.

Here is an example of part of a stream map. The streams in this generated map already look way more similar to what you would find in actual ranked maps.

https://user-images.githubusercontent.com/47356629/184454735-e9cbbecd-18b5-4202-9eb4-e3652b666eae.mp4

I also did a few code readability improvements. I extracted two conditions to local methods (which was previously suggested by @bdach https://github.com/ppy/osu/pull/18769#discussion_r901106782) and added comments that explain the different offsets and how/when they are (re-)calculated.

Pasi4K5 committed:
Adjust angle and `sectionOffset` calculations

Pasi4K5 committed:
Improve code quality

peppy committed:
Merge branch 'master' into enhance-randomisation

peppy committed:
Make `random` implicitly null to avoid potential incorrect behaviour in `randomBool`

peppy committed:
Refactor `IsHitObjectOnBeat` to be understandable

peppy commented:
Let's get this in. I think it feels better than what we have, and the code reads well enough.
